core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala

```diff
@@ -308,6 +309,14 @@ class HadoopRDD[K, V](
     // Do nothing. Hadoop RDD should not be checkpointed.
   }
 
+  override def persist(storageLevel: StorageLevel): this.type = {
+    if (storageLevel.deserialized) {
+      throw new SparkException("Can't cache HadoopRDDs as deserialized objects because Hadoop's" +
```

Comment:
Also the error message might include the workaround; right now, it seems somewhat final.